--- u-boot-linaro-stable-2011.10//boards.cfg	2011-10-24 14:31:53.000000000 +0800
+++ opencsbc-u-boot-mini210-tiny210-last//boards.cfg	2012-12-02 14:01:21.000000000 +0800
@@ -181,6 +181,8 @@
 omap4_panda                  arm         armv7       panda               ti             omap4
 omap4_sdp4430                arm         armv7       sdp4430             ti             omap4
 s5p_goni                     arm         armv7       goni                samsung        s5pc1xx
+mini210                      arm         armv7       mini210             samsung        s5pc1xx
+tiny210                      arm         armv7       tiny210             samsung        s5pc1xx
 smdkc100                     arm         armv7       smdkc100            samsung        s5pc1xx
 origen			     arm	 armv7	     origen		 samsung	s5pc2xx
 s5pc210_universal            arm         armv7       universal_c210      samsung        s5pc2xx
--- a/Makefile	2011-10-24 14:31:53.000000000 +0800
+++ b/Makefile	2013-01-13 20:16:54.687100753 +0800
@@ -23,7 +23,7 @@
 
 VERSION = 2011
 PATCHLEVEL = 09
-SUBLEVEL =
+SUBLEVEL =OpenWrt
 EXTRAVERSION =
 ifneq "$(SUBLEVEL)" ""
 U_BOOT_VERSION = $(VERSION).$(PATCHLEVEL).$(SUBLEVEL)$(EXTRAVERSION)
