--- u-boot-2011.06/drivers/mmc/Makefile	2011-06-28 04:22:42.000000000 +0800
+++ u-boot-2011.06-tekkaman/drivers/mmc/Makefile	2012-12-17 15:15:02.000000000 +0800
@@ -38,6 +38,7 @@
 COBJS-$(CONFIG_OMAP_HSMMC) += omap_hsmmc.o
 COBJS-$(CONFIG_PXA_MMC) += pxa_mmc.o
 COBJS-$(CONFIG_S5P_MMC) += s5p_mmc.o
+COBJS-$(CONFIG_S3C64X0_MMC) += s3c64x0_mmc.o
 
 COBJS	:= $(COBJS-y)
 SRCS	:= $(COBJS:.o=.c)
--- u-boot-2011.06/drivers/mmc/mmc.c	2011-06-28 04:22:42.000000000 +0800
+++ u-boot-2011.06-tekkaman/drivers/mmc/mmc.c	2012-12-17 15:15:02.000000000 +0800
@@ -165,11 +165,19 @@
 	list_for_each(entry, &mmc_devices) {
 		m = list_entry(entry, struct mmc, link);
 
-		if (m->block_dev.dev == dev_num)
+		if (m->block_dev.dev == dev_num) {
+			if (m->detect_mmc) {
+				if (m->detect_mmc((struct mmc_host *)m->priv) == 0) {
+					printf("No SD/MMC card detected!\n");
+					return NULL;
+				}
+			}
+
 			return m;
+		}
 	}
 
-	printf("MMC Device %d not found\n", dev_num);
+	printf("MMC Host control CH%d not found\n", dev_num);
 
 	return NULL;
 }
@@ -415,6 +423,10 @@
 	mmc->ocr = cmd.response[0];
 
 	mmc->high_capacity = ((mmc->ocr & OCR_HCS) == OCR_HCS);
+
+	if (mmc->high_capacity)	{
+		printf("SD card is SDHC!\n");
+	}
 	mmc->rca = 0;
 
 	return 0;
@@ -478,6 +490,10 @@
 	mmc->ocr = cmd.response[0];
 
 	mmc->high_capacity = ((mmc->ocr & OCR_HCS) == OCR_HCS);
+
+	if (mmc->high_capacity)	{
+		printf("SD card is SDHC!\n");
+	}
 	mmc->rca = 0;
 
 	return 0;
@@ -610,7 +626,7 @@
 	return mmc_send_cmd(mmc, &cmd, &data);
 }
 
-
+//FIXME: there're some problems on SCR read function
 int sd_change_freq(struct mmc *mmc)
 {
 	int err;
@@ -664,12 +680,15 @@
 	switch ((mmc->scr[0] >> 24) & 0xf) {
 		case 0:
 			mmc->version = SD_VERSION_1_0;
+			printf("SD card is SD_VERSION_1_0\n");
 			break;
 		case 1:
 			mmc->version = SD_VERSION_1_10;
+			printf("SD card is SD_VERSION_1_10\n");
 			break;
 		case 2:
 			mmc->version = SD_VERSION_2;
+			printf("SD card is SD_VERSION_2\n");
 			break;
 		default:
 			mmc->version = SD_VERSION_1_0;
@@ -1009,6 +1028,8 @@
 			(mmc->cid[2] >> 24) & 0xf);
 	init_part(&mmc->block_dev);
 
+	printf("SD/MMC size : %d block * %dB = %dMB\n",
+		mmc->block_dev.lba, mmc->read_bl_len, mmc->block_dev.lba/2/1024);
 	return 0;
 }
 
@@ -1129,7 +1150,7 @@
 	list_for_each(entry, &mmc_devices) {
 		m = list_entry(entry, struct mmc, link);
 
-		printf("%s: %d", m->name, m->block_dev.dev);
+		printf("%s: Host controler CH%d", m->name, m->block_dev.dev);
 
 		if (entry->next != &mmc_devices)
 			printf("%c ", separator);
