--- u-boot-2011.06/drivers/video/videomodes.h	2011-06-28 04:22:42.000000000 +0800
+++ u-boot-2011.06-tekkaman/drivers/video/videomodes.h	2012-12-17 15:15:03.000000000 +0800
@@ -23,7 +23,7 @@
 
 
 #ifndef CONFIG_SYS_DEFAULT_VIDEO_MODE
-#define CONFIG_SYS_DEFAULT_VIDEO_MODE	0x301
+#define CONFIG_SYS_DEFAULT_VIDEO_MODE	0x212
 #endif
 
 /* Some mode definitions */
@@ -78,13 +78,17 @@
 #define RES_MODE_960_720	3
 #define RES_MODE_1152x864	4
 #define RES_MODE_1280x1024	5
-#define RES_MODES_COUNT		6
+#define RES_MODE_240x320	6
+#define RES_MODE_480x272	7
+#define RES_MODE_800x480	8
+#define RES_MODES_COUNT		9
 
-#define VESA_MODES_COUNT 19
+#define VESA_MODES_COUNT 	22
 
 extern const struct ctfb_vesa_modes vesa_modes[];
 extern const struct ctfb_res_modes res_mode_init[];
 
+int video_get_param_len (char *start, char sep);
 int video_get_params (struct ctfb_res_modes *pPar, char *penv);
 
 int video_get_video_mode(unsigned int *xres, unsigned int *yres,
--- u-boot-2011.06/drivers/video/videomodes.c	2011-06-28 04:22:42.000000000 +0800
+++ u-boot-2011.06-tekkaman/drivers/video/videomodes.c	2012-12-17 15:15:03.000000000 +0800
@@ -98,6 +98,9 @@
 	{0x319, RES_MODE_1280x1024, 15},
 	{0x31A, RES_MODE_1280x1024, 16},
 	{0x31B, RES_MODE_1280x1024, 24},
+	{0x211, RES_MODE_240x320, 16},
+	{0x212, RES_MODE_480x272, 16},
+	{0x213, RES_MODE_800x480, 16},
 };
 const struct ctfb_res_modes res_mode_init[RES_MODES_COUNT] = {
 	/* x	 y pixclk   le	ri  up	lo   hs vs  s  vmode */
@@ -107,15 +110,19 @@
 	{960, 720, 13100, 160, 40, 32, 8, 80, 4, 0, FB_VMODE_NONINTERLACED},
 	{1152, 864, 12004, 200, 64, 32, 16, 80, 4, 0, FB_VMODE_NONINTERLACED},
 	{1280, 1024, 9090, 200, 48, 26, 1, 184, 3, 0, FB_VMODE_NONINTERLACED},
+	{240,   320,    90000, 1,     4,      1,     1,     30,     4, 0, FB_VMODE_NONINTERLACED},
+	{480,   272,    75000, 2,     3,      1,      1,     40,     1, 0, FB_VMODE_NONINTERLACED},
+	{800,   480,    50000, 2,     2,      2,      2,     41,     4, 0, FB_VMODE_NONINTERLACED},
 };
 
+
 /************************************************************************
  * Get Parameters for the video mode:
  */
 /*********************************************************************
  * returns the length to the next seperator
  */
-static int
+int
 video_get_param_len (char *start, char sep)
 {
 	int i = 0;
