--- u-boot-2010.03/net/tftp.c.orig	2012-12-12 22:05:17.331790521 +0800
+++ u-boot-2010.03/net/tftp.c	2012-12-12 22:05:27.295790527 +0800
@@ -11,7 +11,7 @@
 #include "bootp.h"
 
 #define WELL_KNOWN_PORT	69		/* Well known TFTP port #		*/
-#define TIMEOUT		5000UL		/* Millisecs to timeout for lost pkt */
+#define TIMEOUT		(10*5000UL)		/* Millisecs to timeout for lost pkt */
 #ifndef	CONFIG_NET_RETRY_COUNT
 # define TIMEOUT_COUNT	10		/* # of timeouts before giving up  */
 #else
