--- u-boot-2010.03//include/asm-arm/u-boot-arm.h	2010-04-01 05:54:39.000000000 +0800
+++ u-boot-2010.03-24xx//include/asm-arm/u-boot-arm.h	2012-09-05 15:06:15.710689762 +0800
@@ -35,6 +35,7 @@
 extern ulong _bss_end;		/* BSS end */
 extern ulong IRQ_STACK_START;	/* top of IRQ stack */
 extern ulong FIQ_STACK_START;	/* top of FIQ stack */
+extern ulong bBootFrmNORFlash;	/* bootflag */
 
 /* cpu/.../cpu.c */
 int	cpu_init(void);
diff -Nur u-boot-2010.03//include/linux/mtd/mtd.h u-boot-2010.03-24xx//include/linux/mtd/mtd.h
--- u-boot-2010.03//include/linux/mtd/mtd.h	2010-04-01 05:54:39.000000000 +0800
+++ u-boot-2010.03-24xx//include/linux/mtd/mtd.h	2012-09-05 15:06:15.710689762 +0800
@@ -10,6 +10,7 @@
 #include <linux/types.h>
 #include <div64.h>
 #include <linux/mtd/mtd-abi.h>
+#include <config.h>
 
 #define MTD_CHAR_MAJOR 90
 #define MTD_BLOCK_MAJOR 31
@@ -129,6 +130,12 @@
 	 */
 	u_int32_t writesize;
 
+#if defined(ENABLE_CMD_NAND_YAFFS)
+		/*Thanks for hugerat's code*/
+	u_char rw_oob;
+	u_char skipfirstblk;
+#endif
+
 	u_int32_t oobsize;   /* Amount of OOB data per block (e.g. 16) */
 	u_int32_t oobavail;  /* Available OOB bytes per block */
 
--- u-boot-2010.03//include/mmc.h	2010-04-01 05:54:39.000000000 +0800
+++ u-boot-2010.03-24xx//include/mmc.h	2012-09-05 15:06:15.710689762 +0800
@@ -169,7 +169,7 @@
 #define MMC_RSP_R6      (MMC_RSP_PRESENT|MMC_RSP_CRC|MMC_RSP_OPCODE)
 #define MMC_RSP_R7      (MMC_RSP_PRESENT|MMC_RSP_CRC|MMC_RSP_OPCODE)
 
-
+#if 0
 struct mmc_cid {
 	unsigned long psn;
 	unsigned short oid;
@@ -218,7 +218,7 @@
 	u8	crc:7;
 	u8	one:1;
 };
-
+#endif
 struct mmc_cmd {
 	ushort cmdidx;
 	uint resp_type;
@@ -268,8 +268,10 @@
 
 int mmc_register(struct mmc *mmc);
 int mmc_initialize(bd_t *bis);
-int mmc_init(struct mmc *mmc);
-int mmc_read(struct mmc *mmc, u64 src, uchar *dst, int size);
+//int mmc_init(struct mmc *mmc);
+//int mmc_read(struct mmc *mmc, u64 src, uchar *dst, int size);
+int mmc_init(int verbose);
+int mmc_read(ulong src, uchar *dst, int size);
 struct mmc *find_mmc_device(int dev_num);
 void print_mmc_devices(char separator);
 int board_mmc_getcd(u8 *cd, struct mmc *mmc);
--- u-boot-2010.03//include/part.h	2010-04-01 05:54:39.000000000 +0800
+++ u-boot-2010.03-24xx//include/part.h	2012-09-05 15:06:15.710689762 +0800
@@ -62,6 +62,7 @@
 #define IF_TYPE_MMC		6
 #define IF_TYPE_SD		7
 #define IF_TYPE_SATA		8
+#define IF_TYPE_SDHC		9
 
 /* Part types */
 #define PART_TYPE_UNKNOWN	0x00
--- u-boot-2010.03//include/serial.h	2010-04-01 05:54:39.000000000 +0800
+++ u-boot-2010.03-24xx//include/serial.h	2012-09-05 15:06:15.710689762 +0800
@@ -37,7 +37,7 @@
 
 #endif
 
-#if defined(CONFIG_S3C2410)
+#if defined(CONFIG_S3C2410) ||  defined(CONFIG_S3C2440)
 extern struct serial_device s3c24xx_serial0_device;
 extern struct serial_device s3c24xx_serial1_device;
 extern struct serial_device s3c24xx_serial2_device;
